languagetool-core/src/main/java/org/languagetool/rules/spelling/SymSpellRule.java

```diff
@@ -0,0 +1,291 @@
+      }
+      String word = line.substring(0, split);
+      char freqClass = line.charAt(split + 1);
+      int freq = 1 + ((int) freqClass - (int) 'A'); // A - least frequent, Z - most frequent
+      // exact frequencies don't matter, only used as tiebreaker for sorting
+
+      if (!prohibitedWords.contains(word)) {
+        speller.createDictionaryEntry(word, freq, stage);
+        dictWords.incrementAndGet();
+      }
+    });
+    System.out.printf("Loaded %d words from dictionary.%n", dictWords.intValue());
+    speller.commitStaged(stage);
+    long delta = System.currentTimeMillis() - startTime;
+    System.out.printf("Reading dictionaries took %f seconds.%n", (float) delta / 1000.0);
+    return speller;
+  }
+
+  private void initParameters() {
+    if (SuggestionsChanges.getInstance() != null &&
+      SuggestionsChanges.getInstance().getCurrentExperiment() != null) {
+      if (SuggestionsChanges.getInstance().getCurrentExperiment().parameters.get("candidates") != null) {
+        String candidatesParam = (String) SuggestionsChanges.getInstance().getCurrentExperiment()
+          .parameters.get("candidates");
+        verbosity = SymSpell.Verbosity.valueOf(candidatesParam);
+      }
+      if (SuggestionsChanges.getInstance().getCurrentExperiment().parameters.get("editDistance") != null) {
+        editDistance = (Integer) SuggestionsChanges.getInstance().getCurrentExperiment()
+          .parameters.get("editDistance");
+      }
+      if (SuggestionsChanges.isRunningExperiment("SymSpell+NewSuggestionsOrderer")) {
+        orderer = new SuggestionsOrdererFeatureExtractor(language, languageModel);
+      }
+    }
+
+  }
+
+  public SymSpellRule(ResourceBundle messages, Language language, UserConfig userConfig) {
+    this(messages, language, userConfig, Collections.emptyList());
+  }
+
+  public SymSpellRule(ResourceBundle messages, Language language, UserConfig userConfig, List<Language> altLanguages) {
+    this(messages, language, userConfig, altLanguages, null);
+  }
+
+  public SymSpellRule(ResourceBundle messages, Language language, UserConfig userConfig, List<Language> altLanguages, @Nullable LanguageModel languageModel) {
+    super(messages, language, userConfig, altLanguages, languageModel);
+    initParameters();
+    defaultDictSpeller = spellerCache.getUnchecked(language);
+    userDictSpeller = initUserDictSpeller(userConfig);
+  }
+
+  @Override
+  public String getId() {
+    return "SYMSPELL_RULE";
+  }
+
+  @Override
+  public String getDescription() {
+    return "Spell checking rule using SymSpell algorithm";
+  }
+
+  @Override
+  public RuleMatch[] match(AnalyzedSentence sentence) throws IOException {
+    List<RuleMatch> matches = new ArrayList<>();
+    Set<String> ignoredWords = ignoredWordsCache.getUnchecked(language);
+    for (AnalyzedTokenReadings token : sentence.getTokensWithoutWhitespace()) {
+      if (token.isSentenceStart() || token.isSentenceEnd() || token.isNonWord())
```

Comment:
Really token.isSentenceEnd()? Aren't there cases where the sentence ends with a word?